### PR TITLE
Filter out default member attribute for types with indexer. Implement filtering out base type and interfaces.

### PR DESCRIPTION
**Pull request by andriipatsula**
Epic: https://github.com/dotnet/arcade/issues/10291
* Add filtering of `DefaultMember` attributes for types with indexer
* Implement filtering out base type and interfaces.

### CODE REVIEW

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.DotNet.GenAPI
+{
+    internal static class INamedTypeSymbolExtension
+    {
+        public static bool HasIndexer(this INamedTypeSymbol type)
+        {
+            return type.GetMembers().Where(m => m is IPropertySymbol).Select(m => (IPropertySymbol)m).Any(m => m.IsIndexer);
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:13]:
```suggestion
            return type.GetMembers().Any(member => member is IPropertySymbol propertySymbol && propertySymbol.IsIndexer);
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 2]
```diff
@@ -92,5 +95,20 @@ private static ConstructorInitializerSyntax GenerateBaseConstructorInitializer(I
 
             return constructorInitializer;
         }
+
+        private static BaseListSyntax? GetBaseTypeList(this SyntaxGenerator syntaxGenerator,
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:99]: Please add a comment to describe the method's purpose. I assume its for filtering.

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 3]
```diff
@@ -92,5 +95,20 @@ private static ConstructorInitializerSyntax GenerateBaseConstructorInitializer(I
...
+        {
+            List<BaseTypeSyntax> baseTypes = new();
+
+            if (type.TypeKind == TypeKind.Class && type.BaseType != null && symbolFilter.Include(type.BaseType))
+            {
+                baseTypes.Add(SyntaxFactory.SimpleBaseType((TypeSyntax)syntaxGenerator.TypeExpression(type.BaseType)));
+            }
+
+            baseTypes.AddRange(type.Interfaces.Where(symbolFilter.Include).Select(i => SyntaxFactory.SimpleBaseType((TypeSyntax)syntaxGenerator.TypeExpression(i))));
+            return baseTypes.Count == 0 ? null : SyntaxFactory.BaseList(SyntaxFactory.SeparatedList(baseTypes));
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:111]: Style nit. Start with the positive case first and wrap into multiple lines for readability.

```suggestion
            return baseTypes.Count > 0 ?
                SyntaxFactory.BaseList(SyntaxFactory.SeparatedList(baseTypes)) :
                null;
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs [hunk 4]
```diff
@@ -92,5 +95,20 @@ private static ConstructorInitializerSyntax GenerateBaseConstructorInitializer(I
...
+            ISymbolFilter symbolFilter)
+        {
+            List<BaseTypeSyntax> baseTypes = new();
+
+            if (type.TypeKind == TypeKind.Class && type.BaseType != null && symbolFilter.Include(type.BaseType))
+            {
+                baseTypes.Add(SyntaxFactory.SimpleBaseType((TypeSyntax)syntaxGenerator.TypeExpression(type.BaseType)));
+            }
+
+            baseTypes.AddRange(type.Interfaces.Where(symbolFilter.Include).Select(i => SyntaxFactory.SimpleBaseType((TypeSyntax)syntaxGenerator.TypeExpression(i))));
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:110]: Please add a comment here as well. I assume this adds non filtered out interfaces?

**Review by ViktorHofer — APPROVED** [review]
LGTM except a few style nits. Great job!

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs (hunk 1 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs]: yeh, looks better)

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 2 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:100]: added `Gets the list of base class and interfaces for a given symbol <see cref="INamedTypeSymbol"/>.`

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 4 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs:112]: added

File: src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs (hunk 3 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxGeneratorExtensions.cs]: Thanks, updated

**Pull request merged**